crates/red_knot/src/module.rs

```diff
@@ -0,0 +1,708 @@
+#![allow(unreachable_pub)]
+
+use std::fmt::Formatter;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+
+use crate::files::{FileId, Files};
+use crate::FxDashMap;
+use dashmap::mapref::entry::Entry;
+
+/// ID uniquely identifying a module.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleId(u32);
+
+/// A module name, e.g. `foo.bar`.
+///
+/// Always normalized to the absolute form (never a relative module name).
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleName(smol_str::SmolStr);
+
+#[allow(unused)]
+impl ModuleName {
+    pub fn new(name: &str) -> Self {
+        debug_assert!(!name.is_empty());
+
+        Self(smol_str::SmolStr::new(name))
+    }
+
+    pub fn relative(_dots: u32, name: &str, _to: &Path) -> Self {
+        // FIXME: Take `to` and `dots` into account.
+        Self(smol_str::SmolStr::new(name))
+    }
+
+    pub fn from_relative_path(path: &Path) -> Option<Self> {
```

Comment:
I don't see us throwing away any information, considering that this method doesn't do any Io. The idea here is that we get a full qualified name that we can then throw into the resolve function (that retrieves all information)

Reply:
Ah, yeah, I was looking at this method in isolation; I see the only place it's actually used is in `resolve_path`, where the very next step is to resolve that module name back to a path (and make sure it actually resolves to the right path). So I agree, we will get all that information from resolving, and that's where we should get it from.

I would still prefer for this to be a private method of `ModuleResolver` rather than a public constructor of `ModuleName`, because I think it's kind of important to ensure that it only be used in the context of `resolve`, as just described. We don't want some code in the future constructing `ModuleName` using `ModuleName::from_relative_path` and assuming that means it has a correct module name to path correspondence. But this is more a future thing for robustness, not an important prototype consideration.

Reply:
Yeah I don't mind making it private :)